### ISSUE TITLE
Need to encode each path segment

### ISSUE THREAD

**Issue by Ryan-McMillan** [tivac/aws-sig#12]
Was seeing some signing errors and narrowed it down to path segments needing to be encoded.

It looks like adding ".map(encodeURIComponent)" in the path function solves the issue:

```
function path (ref) {
        var url = ref.url;

        return url.pathname
        .replace(multipleSlashesRegex, "/")
        .split("/")
        .reduce(function (prev, curr) {
            if(curr === "..") {
                prev.pop();

                return prev;
            }

            if(curr === ".") {
                return prev;
            }

            prev.push(curr);

            return prev;
        }, [])
        .map(encodeURIComponent)
        .join("/");
}
```

References: (ctrl + f "segment")
https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
https://docs.aws.amazon.com/AlexaWebInfoService/latest/CalculatingSignatures.html

**Comment by tivac** [comment]
Need to be sure that this handles url encoded values as part of the path.

Here's an example of the raw value:

```
arn:aws:thing:us-west-2::session/fleet-cd38eef3-f782-487d-a9cd-6bc188d4a9c3/e16d3f92-2d22-4844-995c-4b52bd98fcac
```
And here's an example of what it would look like in a URL:

```
https://aws.amazon.com/arn%3Aaws%3Athing%3Aus-west-2%3A%3Asession%2Ffleet-cd38eef3-f782-487d-a9cd-6bc188d4a9c3%2Fe16d3f92-2d22-4844-995c-4b52bd98fcac
```

Here's what needs to happen:

- Remove decode call
- Change test to ensure that the weird values are url-encoded before passing to `sign()`
- Document in the README that URLs should be encoded **before** passing to `aws-sig`

**Comment by tivac** [comment]
This change started failing some of the official AWS sigv4 test suite examples and I was super concerned I was misunderstanding something.

Nope. Turns out the official test suite has **bad examples** in it. Oy vey.

- https://github.com/FantasticFiasco/aws-signature-version-4#the-pledge
- https://github.com/aws/aws-sdk-js/issues/853#issuecomment-167843405

So, given that I've grabbed the latest version of the sigv4 test suite and I'm gonna stare at that chart for a while and see if I can figure out wtf to do here.
